src/main/java/teammates/common/util/Url.java

```diff
@@ -1,15 +1,17 @@
 package teammates.common.util;
 
-public class Url {
+/**
+ * The Url class represents a URL string.
+ * It provides methods to manipulate the URL string and extract values from it.
+ */
+public abstract class Url {
```

Comment:
I don't see the need for having setters. In fact I could make everything from base URL to its path `final` (I've done half of it) and only make the query part editable (not even fully editable; can be appended, but cannot be overwritten). Java's URL doesn't have setters either.
Also, the relative path needs a leading slash. That's how Java's URL work, and that's how it's done in TEAMMATES.